### PR TITLE
Table indices tidying: class structure and repr's

### PR DESCRIPTION
**Pull request by taldcroft**
### Description

This pull request was the start of a fix for #11332, but I found that I needed to clean up the basic code structure to try making sense of things. In particular there was a pretty weird construct where the `Index` class initializer was actually returning an instance of `SlicedIndex`. Combining this with the entirely inscrutable `repr()` of various bits has always made table indices really confusing to me.

## Example repr outputs
```
from astropy import table
from astropy.table.table_helpers import simple_table

for engine in table.BST, table.SortedArray, table.SCEngine:
    t = simple_table(size=100)
    t.add_index('a', engine=engine)
    print()
    print(engine)
    print(t.indices)
```
gives:
```
<class 'astropy.table.bst.BST'>
[<SlicedIndex original=True index=<Index columns=('a',) data=<BST>>>]

<class 'astropy.table.sorted_array.SortedArray'>
[<SlicedIndex original=True index=<Index columns=('a',) data=<SortedArray length=100>
 a  rows
--- ----
  1    0
  2    1
  3    2
  4    3
  5    4
...  ...
 95   94
 96   95
 97   96
 98   97
 99   98
100   99
Length = 100 rows>>]

<class 'astropy.table.soco.SCEngine'>
[<SlicedIndex original=True index=<Index columns=('a',) data=<SCEngine nodes=Node((1,), 0), Node((2,), 1), Node((3,), 2), ..., Node((98,), 97), Node((99,), 98), Node((100,), 99)>>>]
```

### CODE REVIEW

File: astropy/table/bst.py [hunk 1]
```diff
@@ -443,12 +443,10 @@ def _same_prefix(self, val, node, lst):
         return lst
 
     def __str__(self):
```
Comment (mhvk) [inline astropy/table/bst.py:445]: I think you can just remove `__str__` if you want it to be the same as `__repr__`

File: astropy/table/bst.py [hunk 2]
```diff
@@ -443,12 +443,10 @@ def _same_prefix(self, val, node, lst):
...
-        if self.root is None:
-            return 'Empty'
-        return self._print(self.root, 0)
+        return repr(self)
 
     def __repr__(self):
-        return str(self)
+        return f'<{self.__class__.__name__}>'
```
Comment (mhvk) [inline astropy/table/bst.py:449]: Maybe just remove `__repr__` as well if we're not using any information? The class name will be given by default...

File: astropy/table/index.py [hunk 3]
```diff
@@ -384,10 +371,10 @@ def __getitem__(self, item):
         return SlicedIndex(self, item)
 
     def __str__(self):
```
Comment (mhvk) [inline astropy/table/index.py:373]: As above, just remove.

File: astropy/table/index.py [hunk 4]
```diff
@@ -384,10 +371,10 @@ def __getitem__(self, item):
...
-        return str(self.data)
+        return repr(self)
 
     def __repr__(self):
-        return str(self)
+        return f'<{self.__class__.__name__} data={self.data}>'
```
Comment (mhvk) [inline astropy/table/index.py:377]: Yes to information!

File: astropy/table/index.py [hunk 5]
```diff
@@ -565,9 +552,14 @@ def sort(self):
...
-        if self.original:
-            return repr(self.index)
-        return f'Index slice {self.start, self.stop, self.step} of\n{self.index}'
+        col_names = tuple(col.info.name for col in self.index.columns)
+        if len(col_names) == 1:
+            name_str = f'column={col_names[0]!r}'
+        else:
+            name_str = f'columns={col_names}'
+        slice_str = '' if self.original else f' slice={self.start}:{self.stop}:{self.step}'
+        return (f'<{self.__class__.__name__} {name_str} original={self.original}{slice_str}'
```
Comment (mhvk) [inline astropy/table/index.py:561]: I'd be tempted to be closer to what `__init__` takes, but this is definitely a huge improvement.

File: astropy/table/index.py [hunk 6]
```diff
@@ -594,11 +586,14 @@ def get_slice(self, col_slice, item):
         '''
         from .table import Table
         if len(self.columns) == 1:
-            return Index([col_slice], engine=self.data.__class__)
+            index = Index([col_slice], engine=self.data.__class__)
+            return SlicedIndex(index, slice(0, 0, None), original=True)
```
Comment (mhvk) [inline astropy/table/index.py:590]: Can't really judge this, but it certainly seems more logical. Maybe even more logical would be `self.__class__(index, ...)`

File: astropy/table/index.py [hunk 7]
```diff
@@ -594,11 +586,14 @@ def get_slice(self, col_slice, item):
...
+
         t = Table(self.columns, copy_indices=False)
         with t.index_mode('discard_on_copy'):
             new_cols = t[item].columns.values()
-        return Index(new_cols, engine=self.data.__class__)
+        index = Index(new_cols, engine=self.data.__class__)
+        return SlicedIndex(index, slice(0, 0, None), original=True)
```
Comment (mhvk) [inline astropy/table/index.py:596]: idem

File: astropy/table/table.py [hunk 8]
```diff
@@ -1011,11 +1011,13 @@ def add_index(self, colnames, engine=None, unique=False):
...
 
+        is_primary = not self.indices
         index = Index(columns, engine=engine, unique=unique)
-        if not self.indices:
+        sliced_index = SlicedIndex(index, slice(0, 0, None), original=True)
+        if is_primary:
             self.primary_key = colnames
         for col in columns:
-            col.info.indices.append(index)
+            col.info.indices.append(sliced_index)
```
Comment (mhvk) [inline astropy/table/table.py:1020]: This part is hard to judge without actually understanding the code...

**Review by mhvk — COMMENTED** [review]
Bit of a drive-by review. Overall, yes, yes, let's have meaningful `__repr__`! And slowly changing things which seem odd is definitely good. Some small in-line comments.

File: astropy/table/bst.py (hunk 1 above)
Comment (taldcroft) [reply astropy/table/bst.py]: Done

File: astropy/table/bst.py (hunk 2 above)
Comment (taldcroft) [reply astropy/table/bst.py:446]: I don't like the default repr that has (in this case) useless object address info. This current version makes the repr consistent with the others.

File: astropy/table/index.py (hunk 3 above)
Comment (taldcroft) [reply astropy/table/index.py]: Done.

File: astropy/table/bst.py (hunk 2 above)
Comment (taldcroft) [reply astropy/table/bst.py:446]: Also, basically the `BST` class should not be used by anybody in practice. I updated the docs accordingly.

File: astropy/table/index.py (hunk 5 above)
Comment (taldcroft) [reply astropy/table/index.py]: Good point and I moved the `columns` into the `Index` class where it actually lives. So this moves closer to what's in `__init__`. In the end these classes are not really ever instantiated by users so it's worth being pragmatic about the repr. 

The question is whether the new repr conveys what is needed by users to make (some) sense of this. The previous version was always completely baffling to me and I have repeatedly spent time over the years trying to drill down into `indices` and being confused.

File: astropy/table/index.py (hunk 6 above)
Comment (taldcroft) [reply astropy/table/index.py]: Done.

File: astropy/table/index.py (hunk 7 above)
Comment (taldcroft) [reply astropy/table/index.py]: Done.

File: astropy/table/table.py (hunk 8 above)
Comment (taldcroft) [reply astropy/table/table.py:1020]: The old code for `Index.__new__()` was (effectively):
```
    def __new__(cls, *args, **kwargs):
        self = super().__new__(cls)
        self.__init__(*args, **kwargs)
        return SlicedIndex(self, slice(0, 0, None), original=True)
```
I won't say that I understand all the code, but I do believe that the new code here is equivalent to what the old code was doing.  🤞 that the testing is good enough.

**Comment by taldcroft** [comment]
@mvhk - thanks for the quick review. I think I've implemented all but one of your suggestions and done the rest of what's needed for merging this PR (a new test and docs update).

File: astropy/table/bst.py (hunk 2 above)
Comment (mhvk) [reply astropy/table/bst.py:446]: OK, fine to leave it as you have it.

File: astropy/table/index.py [hunk 9]
```diff
@@ -594,11 +577,14 @@ def get_slice(self, col_slice, item):
...
         if len(self.columns) == 1:
-            return Index([col_slice], engine=self.data.__class__)
+            index = Index([col_slice], engine=self.data.__class__)
+            return self.__class__(index, slice(0, 0, None), original=True)
+
         t = Table(self.columns, copy_indices=False)
         with t.index_mode('discard_on_copy'):
             new_cols = t[item].columns.values()
-        return Index(new_cols, engine=self.data.__class__)
+        index = Index(new_cols, engine=self.data.__class__)
```
Comment (mhvk) [inline astropy/table/index.py:586]: Bit worrying that these lines are not covered...

File: astropy/table/index.py [hunk 10]
```diff
@@ -537,40 +524,36 @@ def replace(self, row, col, val):
         if not self._frozen:
             self.index.replace(self.orig_coords(row), col, val)
 
-    def copy(self):
+    def get_index_or_copy(self):
```
Comment (mhvk) [inline astropy/table/index.py:527]: Not for now, but this behaviour seems truly bizarre. At least with the name change it is slightly less confusing, but really no-one would expect that this influences the instance itself!

**Review by mhvk — APPROVED** [review]
Am OK with relying on the tests here, since this is clearly an improvement. But I wish I understood better how it is supposed to work...

**Pull request merged**